systemd/nslogin/nslogin.c

```diff
@@ -0,0 +1,315 @@
+                                           "org.freedesktop.systemd1.Manager", "SystemState", &err, &msg);
+    if (busOk < 0) {
+        fprintf(stderr, "Bus error %s", err.message);
+        return false;
+    }
+
+    if (msg == NULL) {
+        return false;
+    }
+    // We won't get the message "Failed to connect to bus: No such file or directory" at this point because there is a
+    // bus available at this point.
+    bool starting = (strncmp(msg, "initializing", 12) == 0 || strncmp(msg, "starting", 8) == 0);
+    free(msg);
+    return !starting;
+}
+
+/// Returns 0 if the basename of the symlink target matches the expected name up to length.
+static int symlink_basename_cmp(const char *symlink, const char *name, size_t length) {
+    char target[PATH_MAX];
+    ssize_t len = readlink(symlink, target, PATH_MAX);
+    if (len == -1) {
+        // This condition becomes really common if attempting to find the PID by trial-and-error.
+        // Thus, printing this error condition would become just noise.
+        return -5;
+    }
+    target[len] = '\0';
+    char *targetBasename = strrchr(target, '/');
+    if (targetBasename == NULL) {
+        perror("Target base name error");
+        return -5;
+    }
+    targetBasename += 1;
+    return strncmp(targetBasename, name, length);
+}
+
+static bool match_owner(const char *path, uid_t uid, gid_t gid) {
+    struct stat fileStat;
+    if (stat(path, &fileStat) != 0) {
+        char msg[96];
+        snprintf(msg, 96, "Failed to stat %s", path);
+        perror(msg);
+        return false;
+    }
+
+    return uid == fileStat.st_uid && gid == fileStat.st_gid;
+}
+
+pid_t find_systemd(void) {
+    struct procInfo {
+        const char *basename;
+        size_t basenameSize;
+        uid_t owner;
+    };
+
+    struct procInfo systemd = {"systemd", 7, 0};
+    char exeLinkPath[16] = {'\0'};
+    for (int16_t pidCandidate = 2; pidCandidate < INT16_MAX; pidCandidate++) {
+        snprintf(exeLinkPath, 16, "/proc/%" SCNd16 "/exe", pidCandidate);
+        int res = symlink_basename_cmp(exeLinkPath, systemd.basename, systemd.basenameSize);
+        if (res != 0) {
+            continue;
+        }
+        if (!match_owner(exeLinkPath, systemd.owner, systemd.owner)) {
+            continue;
+        }
+        return (pid_t)pidCandidate;
+    }
+    return 0;
+}
+
+bool wait_on_systemd(void) {
+    sd_bus *bus;
+    bool systemdStarted = true;
+    int waitCounts = 0;
+    int busOk = -1;
+    const unsigned int timeoutUs = 500000;
+    const int loopMax = 19;
+    for (busOk = sd_bus_default_system(&bus); busOk < 0 && waitCounts < loopMax; waitCounts++) {
+        usleep(timeoutUs); // 500 ms.
+        busOk = sd_bus_default_system(&bus);
+    }
+    // If a timeout had occurred before and bus was still unavailable, the following loop will not even run.
+    for (; busOk >= 0 && waitCounts <= loopMax && !(systemdStarted = is_systemd_startup_complete(bus)); waitCounts++) {
```

Comment:
Maybe add a comment to explicitely state we don’t restart the timeout and that it’s global to both operations? (availibity on bus AND started)